✨amp-consent v2: Fullscreen API Support & UX

torch2424 opened the pull request:
relates to #18828

* Added a JS API for `.postMessage` on the parent frame to request fullscreen, and exit fullscreen.
* Added CSS to support a smoot transition between consent frame states. Using CSS3 Transisitions, and CSS3 3DTransforms (For GPU Acceleration).
* Fixed Linting Warnings for Error Messages
* Updated `/test/manual/diy-consent.html`

Console Errors were there before this PR, as they manually add a top to the consent element. Perhaps @zhouyx would have context on why that is 😄 

# Example

**After PR Comments**

![ampconsentfullscreen](https://user-images.githubusercontent.com/1448289/47882623-62d35100-dde7-11e8-983e-f06091b01dd6.gif)

zhouyx commented on extensions/amp-consent/0.1/amp-consent.js:
```diff
@@ -177,10 +179,14 @@ export class AmpConsent extends AMP.BaseElement {
+        () => this.handleAction_(ACTION_TYPE.REJECT));
     this.registerAction('dismiss',
         () => this.handleAction_(ACTION_TYPE.DISMISS));
+    this.registerAction('dismiss',
```
'dismiss' and enter full_screen? 

As far as I tell, the API should live within the the consent-ui.js along with the `consent-ui-ready` API. Could you explain why we need this?

zhouyx commented on extensions/amp-consent/0.1/amp-consent.js:
```diff
@@ -315,6 +321,23 @@ export class AmpConsent extends AMP.BaseElement {
+      return;
+    }
+
+    if (action == ACTION_TYPE.EXIT_FULLSCREEN) {
```
I am not expecting a `EXIT_FULLSCREEN` action. The user flow is as following

PROMPT -> ENTER_FULLSCREEN -> USER INTERACTION & HIDE PROMPT

torch2424 commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -174,6 +182,47 @@ export class ConsentUI {
+    this.isFullscreen_ = false;
+
+    // Set a timeout to remove the fullscreen exit class
+    // This timeout should match our CSS animation speed.
```
Fixed the comment in my current local branch.

So the class removal is required to stop any jumping of the iframe inside the `amp-consent`, it looks really bad. If you look at the CSS, I couldn't think of another way to transition the height, with graphics accelerated CSS, and look nice without the `-exit` class.

I decided on the `transitionend`, because using a setTimeout just seemed incorrect and hack-y.  I didn't consider a user adding a transition themselves. But I have an idea, the [`transitionend event`](https://developer.mozilla.org/en-US/docs/Web/Events/transitionend) expsoses a property, perhaps we can use that in order to ensure that it is our transition, and then act accordingly?

Also, since this is our agreed UX usage of the component, perhaps it would be good to use some `!important` for the animations? Let me know what you think 😄

torch2424 replied:
Also, talked with another UXE, showed them my approach to the problem, and agreed this is probably the best/good way of doing this. We couldn't come up with another CSS solution to achieve this same effect.

torch2424 commented:
@lannka @zhouyx 

Addressed all comments. Also, note in the Original PR post, I added a new gif with some fixes to the transition animation. It now transitions at the end as well 😄

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -31,6 +31,17 @@ import {toggle} from '../../../src/style';
 
+// Classes for consent UI
+const consentUiClasses = {
+  fullscreen: 'i-amphtml-consent-ui-fullscreen',
```
Why we need an object? Also I don't know if we are planning to move to JsonObject, I'm waiting for Erwin's response on this.

torch2424 replied:
I moved this into a class, per @lannka 's suggestion of moving the classes into constants so that we have one place to change it.

I also felt like `consentUiClasses.fullscreen`, would make more sense than `CONSENT_UI_CLASSES_FULLSCREEN`, and mantaining a bunch of individual variables.

But yes, I can set this as JSONObject type 😄

torch2424 replied:
Bracket notiation with Strings

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -235,21 +284,47 @@ export class ConsentUI {
     classList.remove('loading');
-    classList.add('i-amphtml-consent-ui-in');
-    classList.add('consent-iframe-active');
+    this.vsync_.mutate(() => {
```
nit: can switch to `this.baseInstance_.mutateElement(() => {})` instead.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -235,21 +284,47 @@ export class ConsentUI {
+   * @param {Function} callback
    */
-  resetIframe_() {
+  resetIframe_(callback) {
```
can we not use callback : )

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -235,21 +284,47 @@ export class ConsentUI {
+    } else {
+      this.win_.removeEventListener('message', this.boundHandleIframeMessages_);
+      classList.remove(consentUiClasses.uiIn);
+      this.getTransformTransitionEndPromise_().then(() => {
```
Didn't fully get this, I don't think we need transition when to hide iframe.

torch2424 replied:
Let's add a comment here to get a UX review

torch2424 commented on extensions/amp-consent/0.1/amp-consent.css:
```diff
@@ -17,12 +17,21 @@
   /* Fixed to make position independent of page other elements. */
   position: fixed;
-  bottom: 0;
+  top: 0;
```
I personally find it easier to reason about the code if we calculate everything starting from the top. If we do `bottom: 0`, then we have to introduce negative y translation, which I think is slightly more confusing

zhouyx commented on extensions/amp-consent/0.1/amp-consent.js:
```diff
@@ -670,7 +683,7 @@ export class AmpConsent extends AMP.BaseElement {
             //    display and display. Add one more check before display
             // 2. Should not schedule display with DISMISSED UNKNOWN state
-            this.scheduleDisplay_(instanceId);
+            return this.scheduleDisplay_(instanceId);
```
Why we have the return here? I don't see the change to `this.scheduleDisplay_` return value.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -153,27 +172,50 @@ export class ConsentUI {
-    toggle(this.ui_, false);
-    if (this.isCreatedIframe_) {
-      this.resetIframe_();
+    if (this.ui_) {
```
We've already check `if(!this.us_)`. No need for the check here. 
If it is for type check, use `assertElement()` instead.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -269,12 +345,19 @@ export class ConsentUI {
     }
+
+    if (data['type'] === 'consent-ui-enter-fullscreen') {
+      this.vsync_.mutate(() => {
```
nit: This is my bad of using `vsync.mutate` 😄  use `mutateElement` instead.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -269,12 +345,19 @@ export class ConsentUI {
+      this.vsync_.mutate(() => {
+        this.enterFullscreen_();
+      });
+      return;
```
nit: remove

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -232,24 +274,52 @@ export class ConsentUI {
-    removeElement(dev().assertElement(this.ui_));
+
+    if (this.isFullscreen_) {
+      this.win_.removeEventListener('message', this.boundHandleIframeMessages_);
```
Can we not do if/else, but always set `this.isFullscreen_` to false, remove classLists, eventListener and elements.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -153,27 +172,50 @@ export class ConsentUI {
    * Hide the UI
    */
   hide() {
+
```
Since you remove the vsync wrapper around `hide`, may want to add it here.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -269,12 +345,19 @@ export class ConsentUI {
       this.iframeReady_.resolve();
     }
+
+    if (data['type'] === 'consent-ui-enter-fullscreen') {
```
what happens if `consent-ui-enter-fullscreen` is received before `consent-ui-ready`?

zhouyx commented on extensions/amp-consent/0.1/test/test-consent-ui.js:
```diff
@@ -49,10 +49,16 @@ describes.realWin('consent-ui', {
-        addToFixedLayer: () => {},
-        removeFromFixedLayer: () => {},
-      };
+      getViewport: () => {
```
There's no test coverage for handling received messages. Mostly because I didn't put one last time : ) Let's create a seperate PR to add them 😄

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -269,12 +341,23 @@ export class ConsentUI {
       return;
     }
 
     if (data['type'] == 'consent-ui-ready') {
```
nit: since you're using `===` here, let's switch to triple equal here as well.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -269,12 +341,23 @@ export class ConsentUI {
+    if (data['type'] === 'consent-ui-enter-fullscreen') {
+
+      if (!this.isIframeVisible_) {
+        return;
```
Do you think we should tell the iframe if the `enter-fullscreen` succeed or not? Please add a TODO and we can address that once we refactor the message type.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -232,23 +277,44 @@ export class ConsentUI {
-    classList.remove('i-amphtml-consent-ui-in');
+
+    // Remove the iframe active to go back to our normal height
     classList.remove('consent-iframe-active');
```
nit: `consentUiClasses.iframeActive`

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -31,6 +31,16 @@ import {toggle} from '../../../src/style';
+// Classes for consent UI
+const consentUiClasses = {
+  fullscreen: 'i-amphtml-consent-ui-fullscreen',
+  iframeActive: 'consent-iframe-active',
```
Let's sync offline on the class name.

torch2424 replied:
I agree. Was trying to follow your example for class names. But I think it would be good to go over them again 😄

torch2424 commented:
Discussed offline with @zhouyx We agreed that we should keep the class names so that in the CMP case they are forced to follow our UI Flow. This is because of several reasons:

1. Publishers that want a custom UI probably want to not use a CMP anyways, and roll their own

2. CMPs may not necessarily have responsive content in their iframe, and allowing publishers to dynamically do anything with their content could make things more difficult for all parties.

3. We can always offer alternative UIs in the CMP config that we offer. From meetings with CMPs, some already have a system where the user picks one of X UIs to use on their website.

Therefore, for our MVP, we will be enforcing the footer bar, to slide in from bottom UI.

torch2424 commented:
@zhouyx Made requested changes, this is now good to go 😄

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -27,10 +27,20 @@ import {
+  iframeFullscreen: 'i-amphtml-consent-ui-iframe-fullscreen',
+  iframeActive: 'i-amphtml-consent-ui-iframe-active',
+  in: 'i-amphtml-consent-ui-in',
+  loading: 'loading',
```
Let's make this `i-amphtml-*` as well.

zhouyx commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -232,23 +274,50 @@ export class ConsentUI {
-    classList.add('i-amphtml-consent-ui-in');
-    classList.add('consent-iframe-active');
+
+    // Remove transition/transform styles added by the fixed layer
```
I'm not sure I understand this, Do we need to reapply the styles afterwards?

torch2424 replied:
In [`src/fixed-layer.js`](https://github.com/ampproject/amphtml/blob/master/src/service/fixed-layer.js#L649), styles are applied to top and transition in order to stop jank and other random bugs with static fixed elements.

However, since we want to animate our element on the fixed layer, we need to undo the styles that are initially applied to stop any weird jank or unexpected behaviors, that was we can add our animations.

Thus, the original styles do not need to be re-applied as the fixed layer service will handle that for us once we remove/add this element in the future.

torch2424 commented:
Currently fighting the tests as they are a bit flaky. Will report when ready

torch2424 commented:
@zhouyx Fixed / replied to your comments. This is now good to go 😄

torch2424 commented:
Fixed some errors after travis failed, this is still good to go 😄

torch2424 commented:
Some tasks to do discussed offline:

Test the following amp pages with the current implementation, specifically, checking the fixed layer styles:

* https://github.com/ampproject/amphtml/blob/master/test/manual/ampUserNotification-ampConsent.html

* https://github.com/ampproject/amphtml/blob/master/test/manual/amp-consent.html

torch2424 commented on extensions/amp-consent/0.1/consent-ui.js:
```diff
@@ -232,23 +274,50 @@ export class ConsentUI {
+     * This will avoid race conditions with the slidein transition.
+     */
+    this.baseInstance_.mutateElement(() => {
+      classList.remove('loading');
```
Nice catch! Thank you! 😄

torch2424 commented:
Yay! Thank you @zhouyx 

Now This just needs approval from you, and @lannka So we can merge this in 😄

zhouyx commented:
Awesome! Thank you @torch2424 Can we add more test coverage during fixit week : )

torch2424 commented:
Forgot to merge this in. Merging...
